// vectra/lib/LocalDocument.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalDocument = void 0;
const fs = __importStar(require("fs/promises"));
const path = __importStar(require("path"));
/**
 * Represents an indexed document stored on disk.
 */
class LocalDocument {
    /**
     * Creates a new `LocalDocument` instance.
     * @param index Parent index that contains the document.
     * @param id ID of the document.
     * @param uri URI of the document.
     */
    constructor(index, id, uri) {
        this._index = index;
        this._id = id;
        this._uri = uri;
    }
    /**
     * Returns the folder path where the document is stored.
     */
    get folderPath() {
        return this._index.folderPath;
    }
    /**
     * Returns the ID of the document.
     */
    get id() {
        return this._id;
    }
    /**
     * Returns the URI of the document.
     */
    get uri() {
        return this._uri;
    }
    /**
     * Returns the length of the document in tokens.
     * @remarks
     * This value will be estimated for documents longer then 40k bytes.
     * @returns Length of the document in tokens.
     */
    getLength() {
        return __awaiter(this, void 0, void 0, function* () {
            const text = yield this.loadText();
            if (text.length <= 40000) {
                return this._index.tokenizer.encode(text).length;
            }
            else {
                return Math.ceil(text.length / 4);
            }
        });
    }
    /**
     * Determines if the document has additional metadata storred on disk.
     * @returns True if the document has metadata; otherwise, false.
     */
    hasMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield fs.access(path.join(this.folderPath, `${this.id}.json`));
                return true;
            }
            catch (err) {
                return false;
            }
        });
    }
    /**
     * Loads the metadata for the document from disk.
     * @returns Metadata for the document.
     */
    loadMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._metadata == undefined) {
                let json;
                try {
                    json = (yield fs.readFile(path.join(this.folderPath, `${this.id}.json`))).toString();
                }
                catch (err) {
                    throw new Error(`Error reading metadata for document "${this.uri}": ${err.toString()}`);
                }
                try {
                    this._metadata = JSON.parse(json);
                }
                catch (err) {
                    throw new Error(`Error parsing metadata for document "${this.uri}": ${err.toString()}`);
                }
            }
            return this._metadata;
        });
    }
    /**
     * Loads the text for the document from disk.
     * @returns Text for the document.
     */
    loadText() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._text == undefined) {
                try {
                    this._text = (yield fs.readFile(path.join(this.folderPath, `${this.id}.txt`))).toString();
                }
                catch (err) {
                    throw new Error(`Error reading text file for document "${this.uri}": ${err.toString()}`);
                }
            }
            return this._text;
        });
    }
}
exports.LocalDocument = LocalDocument;
